File: xmledit.py
from PySide import QtGui
from PySide import QtCore

# taken from ElementLib
def indent(elem, level = 0, tabImpostor = "    "):
    i = "\n" + level * tabImpostor
    if len(elem):
        if not elem.text or not elem.text.strip():
            elem.text = i + tabImpostor
        for e in elem:
            indent(e, level+1)
            if not e.tail or not e.tail.strip():
                e.tail = i + tabImpostor
        if not e.tail or not e.tail.strip():
            e.tail = i
    else:
        if level and (not elem.tail or not elem.tail.strip()):
            elem.tail = i

class XMLSyntaxHighlighter(QtGui.QSyntaxHighlighter):
    def __init__(self, parent=None):
        super(XMLSyntaxHighlighter, self).__init__(parent)
 
        self.highlightingRules = []
 
        # todo: some fail colour highlighting :D please someone change the colours
        keywordFormat = QtGui.QTextCharFormat()
        keywordFormat.setFontWeight(QtGui.QFont.Bold)
        keywordFormat.setForeground(QtCore.Qt.darkCyan)
        
        for pattern in ["\\b?xml\\b", "/>", ">", "</", "<"]:
            self.highlightingRules.append((QtCore.QRegExp(pattern), keywordFormat))
 
        elementNameFormat = QtGui.QTextCharFormat()
        elementNameFormat.setFontWeight(QtGui.QFont.Bold)
        elementNameFormat.setForeground(QtCore.Qt.red)
        
        self.highlightingRules.append((QtCore.QRegExp("\\b[A-Za-z0-9_]+(?=[\s/>])"), elementNameFormat))
 
        attributeKeyFormat = QtGui.QTextCharFormat()
        attributeKeyFormat.setFontItalic(True)
        attributeKeyFormat.setForeground(QtCore.Qt.blue)
        self.highlightingRules.append((QtCore.QRegExp("\\b[A-Za-z0-9_]+(?=\\=)"), attributeKeyFormat))
        
    def highlightBlock(self, text):
        for expression, format in self.highlightingRules:
            index = expression.indexIn(text)

            while index >= 0:
                length = expression.matchedLength()
                self.setFormat(index, length, format)
 
                index = expression.indexIn(text, index + length)
    
class XMLEditWidget(QtGui.QTextEdit):
    def __init__(self):
        super(XMLEditWidget, self).__init__()
        
        self.setAcceptRichText(False)
        self.zoomIn()
        
        # todo: we want dark background!! This doesn't entirely work the way I would want it to,
        #       only areas that have text are with black background :-(
        #self.setTextBackgroundColor(QtGui.QColor(0, 0, 0))
        self.highlighter = XMLSyntaxHighlighter(self.document())
        

File: ceed/src/xmledit.cpp
/*
   CEED - Unified CEGUI asset editor

   Copyright (C) 2011-2017   Martin Preisler <martin@preisler.me>
                             and contributing authors (see AUTHORS file)

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "xmledit.h"

namespace CEED {
namespace xmledit {


void indent(ElementTree::Element *elem, int level, const QString &tabImpostor)
{
#if 0
    QString i = "\n" + tabImpostor.repeated(level);
    if (elem != nullptr) {
        if (elem->text.isEmpty() || elem->text.trimmed().isEmpty())
            elem->text = i + tabImpostor;

        ElementTree::Element* last = nullptr;
        for (ElementTree::Element *e : elem) {
            indent(e, level + 1);
            if (e->tail.isEmpty() || e->tail.strip().isEmpty())
                e->tail = i + tabImpostor;

            last = e;
        }

        if (last->tail.isEmpty() || last->tail.trimmed().isEmpty())
            last->tail = i;
    } else {
        if (level && (elem->tail.isEmpty() || elem->tail.trimmed().isEmpty()))
            elem->tail = i;
    }
#endif
}


XMLSyntaxHighlighter::XMLSyntaxHighlighter(QTextDocument *parent)
    : QSyntaxHighlighter(parent)
{
    // todo: some fail colour highlighting :D please someone change the colours
    QTextCharFormat keywordFormat;
    keywordFormat.setFontWeight(QFont::Bold);
    keywordFormat.setForeground(Qt::darkCyan);

    QStringList patterns = { "\\b?xml\\b", "/>", ">", "</", "<" };
    for (auto pattern : patterns) {
        m_highlightingRules[pattern] = { QRegExp(pattern), keywordFormat };
    }

    QTextCharFormat elementNameFormat;
    elementNameFormat.setFontWeight(QFont::Bold);
    elementNameFormat.setForeground(Qt::red);
    QString pattern = "\\b[A-Za-z0-9_]+(?=[\\s/>])";
    m_highlightingRules[pattern] = { QRegExp(pattern), elementNameFormat };

    QTextCharFormat attributeKeyFormat;
    attributeKeyFormat.setFontItalic(true);
    attributeKeyFormat.setForeground(Qt::blue);
    pattern = "\\b[A-Za-z0-9_]+(?=\\=)";
    m_highlightingRules[pattern] = { QRegExp(pattern), attributeKeyFormat };
}

void XMLSyntaxHighlighter::highlightBlock(const QString &text)
{
    for (auto it = m_highlightingRules.begin(); it != m_highlightingRules.end(); it++) {
        Rule& rule = it.value();
        QRegExp& expression = rule.re;
        int index = expression.indexIn(text);
        while (index >= 0) {
            int length = expression.matchedLength();
            setFormat(index, length, rule.fmt);
            index = expression.indexIn(text, index + length);
        }
    }
}

/////

XMLEditWidget::XMLEditWidget()
    : QTextEdit()
{
    setAcceptRichText(false);
    zoomIn();

    m_highlighter = new XMLSyntaxHighlighter(document());
}

} // namespace xmledit
} // namespace CEED
